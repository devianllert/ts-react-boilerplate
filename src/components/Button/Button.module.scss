@import '../../design/design-system';

.button {
  position: relative;

  display: inline-flex;
  align-items: center;
  justify-content: center;

  height: $button-height;
  min-width: $button-min-width;

  margin: $button-margin;
  padding: $button-padding;

  text-align: center;
  vertical-align: middle;

  font-family: $button-font-family;
  font-size: $button-medium-font-size;
  font-weight: $button-font-weight;

  letter-spacing: $button-letter-spacing;

  border: 0;
  border-radius: $button-border-radius;

  color: var(--button-color);
  background-color: var(--brand-primary);

  text-decoration: $button-text-decoration;
  text-transform: $button-text-transform;

  -webkit-tap-highlight-color: transparent;

  outline: 0;

  appearance: none;

  touch-action: manipulation;
  user-select: none;
  cursor: pointer;

  transition: background-color $button-transition;

  &.small {
    font-size: 12px;
    
    height: $button-small-height;
    min-width: $button-small-min-width;
  }

  &.medium {
    font-size: $button-medium-font-size;
    
    height: $button-medium-height;
    min-width: $button-medium-min-width;
  }

  &.large {
    font-size: $button-medium-font-size;
    
    height: $button-large-height;
    min-width: $button-large-min-width;
  }

  &:hover:before {
    opacity: 0.08;
  }

  &:focus:before {
    opacity: 0.24;
  }

  &:active:before {
    opacity: 0.32;
  }

  &:disabled,
  &.disabled {
    background-color: #CDCDCD;
  }
}

@each $variation, $values in $button-color-variations {
  .#{$variation} {
    color: map-get($values, 'color');

    background-color: map-get($values, 'bg');

    &:before {
      content: "";
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;

      border-radius: inherit;
      
      opacity: 0;

      background-color: #ffffff;

      transition: opacity $button-transition;
    }
  }
}

@each $variation, $values in $button-color-variations {
  .#{$variation}.outlined {
    color: map-get($values, 'bg');

    background-color: transparent;

    border: map-get($values, 'border');

    &:before {
      background-color: currentColor;
    }

    &:disabled,
    &.disabled {
      border-color: #CDCDCD;
    }
  }
}

@each $variation, $values in $button-color-variations {
  .#{$variation}.flat {
    color: map-get($values, 'bg');

    background-color: transparent;

    &:before {
      background-color: currentColor;
    }
  }
}

/* stylelint-disable-next-line no-duplicate-selectors */
.button:disabled,
.button.disabled {
  cursor: not-allowed;

  color: #8E8E8E;

  &:before {
    opacity: 0;
  }
}