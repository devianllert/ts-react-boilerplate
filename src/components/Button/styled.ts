import styled from 'styled-components';
import theme, { ThemeSet } from 'styled-theming';

import {
  PRIMARY,
  SUCCESS,
  WARNING,
  DANGER,
} from '../../design/colors';
import { createTransition, duration } from '../../design/transitions';
import { shape } from '../../design/shape';
import { shadows } from '../../design/shadows';


interface BaseButtonProps {
  disabled?: boolean;
  loading?: boolean;
  outlined?: boolean;
  fullWidth?: boolean;
  flat?: boolean;
  appearence?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning';
  size?: 'small' | 'medium' | 'large';
}

const BUTTON_BACKGROUNDS: Record<string, ThemeSet> = {
  primary: PRIMARY,
  success: SUCCESS,
  warning: WARNING,
  danger: DANGER,
};

const DISABLED_COLOR = theme('mode', {
  light: 'rgba(0, 0, 0, 0.3)',
  dark: 'rgba(255, 255, 255, 0.3)',
});

const DISABLED_BACKGROUND_COLOR = theme('mode', {
  light: 'rgba(0, 0, 0, 0.12)',
  dark: 'rgba(255, 255, 255, 0.12)',
});

export const BaseButton = styled.button<BaseButtonProps>`
  position: relative;

  display: inline-flex;
  align-items: center;
  justify-content: center;

  width: ${({ fullWidth }): string => (fullWidth ? '100%' : 'auto')};
  min-width: 88px;
  height: 36px;

  padding: 0 16px;

  text-align: center;
  vertical-align: middle;

  font-family: "Roboto", sans-serif;
  font-size: 14px;
  font-weight: 500;

  letter-spacing: 1px;

  border: 0;
  border-radius: ${shape.borderRadius}px;

  color: #fff;
  background-color: ${({ appearence = 'primary' }): ThemeSet => BUTTON_BACKGROUNDS[appearence]};

  box-shadow: ${shadows[1]};

  text-decoration: none;
  text-transform: uppercase;

  -webkit-tap-highlight-color: transparent;

  outline: 0;

  appearance: none;

  touch-action: manipulation;
  user-select: none;
  cursor: pointer;

  transition: ${createTransition(['background-color', 'box-shadow', 'border'], {
    duration: duration.short,
  })};

  &:before {
    content: '';

    position: absolute;

    top:0;
    left: 0;

    width: 100%;
    height: 100%;

    background-color: transparent;

    transition: ${createTransition('background-color', { duration: duration.short })};

    user-select: none;
    pointer-events: none;
  }

  &:hover:before {
    background-color: rgba(255, 255, 255, 0.12);
  }

  &:focus:before {
    background-color: rgba(255, 255, 255, 0.18);
  }

  &:active:before {
    background-color: rgba(255, 255, 255, 0.24);
  }

  &:disabled {
    color: ${DISABLED_COLOR};
    background-color: ${DISABLED_BACKGROUND_COLOR};

    box-shadow: none;
  }
`;

export const Label = styled.span`
  width: 100%;
  display: inherit;
  align-items: inherit;
  justify-content: inherit;
`;

export const EndIcon = styled.span`
  display: inherit;
  margin-right: -4px;
  margin-left: 8px;
`;

export const StartIcon = styled.span`
  display: inherit;
  margin-right: 8px;
  margin-left: -4px;
`;
