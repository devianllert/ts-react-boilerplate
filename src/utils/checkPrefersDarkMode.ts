/**
 * Check browser or OS preference for dark mode
 */

const checkPrefersDarkMode = (): boolean => {
  const preferDarkQuery = '(prefers-color-scheme: dark)';

  const mediaQuery = matchMedia(preferDarkQuery);

  return mediaQuery.media === preferDarkQuery;
};

export default checkPrefersDarkMode;
