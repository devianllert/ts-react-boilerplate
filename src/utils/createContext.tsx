import React, {
  createContext,
  useState,
  useContext,
  Provider,
  Dispatch,
  SetStateAction,
  PropsWithChildren,
  ReactNode,
  Context,
} from 'react';

type UpdateType<A> = Dispatch<SetStateAction<A>>;

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export const createCtx = <A extends {} | null>(): readonly [() => A, Provider<A | undefined>] => {
  const ctx = createContext<A | undefined>(undefined);

  const useCtx = (): A => {
    const c = useContext(ctx);

    if (c === undefined) {
      throw new Error('useCtx must be inside a Provider with a value');
    }

    return c;
  };

  return [useCtx, ctx.Provider] as const;
};

export const createCtxWithUpdate = <A extends any>(
  defaultValue: A,
): readonly [
  Context<{ state: A; update: Dispatch<SetStateAction<A>> }>,
  (props: { children?: ReactNode }) => ReactNode,
] => {
  const defaultUpdate: UpdateType<typeof defaultValue> = () => defaultValue;

  const ctx = createContext({
    state: defaultValue,
    update: defaultUpdate,
  });

  const ContextProvider = (props: PropsWithChildren<{}>): ReactNode => {
    const [state, update] = useState(defaultValue);
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <ctx.Provider value={{ state, update }} {...props} />;
  };

  return [ctx, ContextProvider] as const;
};
