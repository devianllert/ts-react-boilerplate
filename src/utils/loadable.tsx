/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/jsx-props-no-spreading */

import React, {
  lazy,
  Suspense,
  ReactElement,
  ComponentType,
} from 'react';

export type ImportComponent = Promise<{ default: any }>;

const loadable = (importFunc: () => ImportComponent, { fallback = null }): ComponentType => {
  const LazyComponent: ComponentType = lazy(importFunc);

  return (props: object): ReactElement => (
    <Suspense fallback={fallback}>
      <LazyComponent {...props} />
    </Suspense>
  );
};

export default loadable;
