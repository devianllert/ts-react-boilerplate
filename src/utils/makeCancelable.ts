/* eslint-disable prefer-promise-reject-errors */
/* https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html */
export interface CancellablePromise<T> {
  promise: Promise<T>;
  cancel(): void;
}

const makeCancelable = <T>(promise: Promise<T>): CancellablePromise<T> => {
  let hasCanceled = false;

  const wrappedPromise = new Promise<T>((resolve, reject) => {
    promise.then(
      (val) => (hasCanceled ? reject({ isCanceled: true }) : resolve(val)),
      (error) => (hasCanceled ? reject({ isCanceled: true }) : reject(error)),
    );
  });

  return {
    promise: wrappedPromise,
    cancel(): void {
      hasCanceled = true;
    },
  };
};

export default makeCancelable;
