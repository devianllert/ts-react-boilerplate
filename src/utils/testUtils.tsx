/* eslint-disable import/prefer-default-export */
import React, { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { i18n as i18Next } from 'i18next';
import { I18nextProvider } from 'react-i18next';
import { render, RenderResult } from '@testing-library/react'; // eslint-disable-line

import configureStore, { EnhancedStore } from '../core/configureStore';
import i18nInstance from '../core/i18n';

export interface RenderProvidersOptions {
  initialState?: {};
  route?: string;
  routerHistory?: MemoryHistory;
  store?: EnhancedStore;
  i18n?: i18Next;
}

export interface RenderProvidersResult extends RenderResult {
  routerHistory: MemoryHistory;
  store: EnhancedStore;
  i18n: i18Next;
}

export type RenderProviders = (ui: ReactElement, options?: RenderProvidersOptions) => RenderProvidersResult;

const renderWithProviders: RenderProviders = (
  ui,
  {
    initialState = {},
    route = '/',
    routerHistory = createMemoryHistory({ initialEntries: [route] }),
    store = configureStore(initialState),
    i18n = i18nInstance,
  }: RenderProvidersOptions = {},
): RenderProvidersResult => ({
  ...render(
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        <Router history={routerHistory}>{ui}</Router>
      </I18nextProvider>
    </Provider>,
  ),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  store,
  routerHistory,
  i18n,
});

export {
  renderWithProviders,
};
