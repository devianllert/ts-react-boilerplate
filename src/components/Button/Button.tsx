import React, { ReactNode, ReactElement, HTMLAttributes } from 'react';
import { Link } from 'react-router-dom';
import { CircularProgress } from 'react-essential-tools';

import * as S from './styled';

interface Props extends HTMLAttributes<HTMLButtonElement | HTMLAnchorElement> {
  /**
   * The content of the button.
   */
  children: ReactNode;
  /**
   * The URL to link to when the button is clicked.
   * If defined, an `a` element will be used as the root node.
   */
  to?: string;
  /**
   * If `true`, the base button will be disabled.
   */
  disabled?: boolean;
  /**
   * If `true`, the base button will have loading indicator.
   */
  loading?: boolean;
  /**
   * If `true`, the button will take up the full width of its container.
   */
  fullWidth?: boolean;
  /**
   * Element placed before the children.
   */
  startIcon?: ReactNode;
  /**
   * Element placed after the children.
   */
  endIcon?: ReactNode;
  type?: 'submit' | 'button' | 'reset';
  /**
   * The variant to use.
   */
  variant?: 'text' | 'outlined' | 'contained';
  /**
   * The color of the component. It supports those theme colors that make sense for this component.
   */
  appearence?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning';
  /**
   * The size of the button.
   * `small` is equivalent to the dense button styling.
   */
  size?: 'small' | 'medium' | 'large';
}

const Button = (props: Props): ReactElement => {
  const {
    children,
    to,
    disabled,
    loading,
    startIcon: startIconProp,
    endIcon: endIconProp,
    type = 'button',
    appearence = 'primary',
    ...otherProps
  } = props;

  const startIcon = startIconProp && <S.StartIcon>{startIconProp}</S.StartIcon>;
  const endIcon = endIconProp && <S.EndIcon>{endIconProp}</S.EndIcon>;

  const content = loading ? (
    <CircularProgress size={24} />
  ) : (
    <S.Label>
      {startIcon}
      {children}
      {endIcon}
    </S.Label>
  );

  if (to && !disabled) {
    return (
      <S.BaseButton
        as={Link}
        appearence={appearence}
        to={to}
        {...otherProps} // eslint-disable-line
      >
        {content}
      </S.BaseButton>
    );
  }

  return (
    <S.BaseButton
      appearence={appearence}
      type={type}
      disabled={disabled}
      {...otherProps} // eslint-disable-line
    >
      {content}
    </S.BaseButton>
  );
};

export default Button;
