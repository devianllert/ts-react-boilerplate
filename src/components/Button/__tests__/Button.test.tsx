import React, { ReactChild } from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';

import { renderWithProviders } from '../../../utils/testUtils';

import '@testing-library/jest-dom/extend-expect';

import Button from '../Button';

describe('<Button />', (): void => {
  afterEach(cleanup);

  const children: ReactChild = 'Button';

  it('should render a children', (): void => {
    const { getByRole } = render(<Button>{children}</Button>);

    expect(getByRole('button')).toHaveTextContent(children);
  });

  it('should render a disabled button if disabled prop is true', (): void => {
    const { getByRole } = render(<Button disabled>{children}</Button>);

    expect(getByRole('button')).toHaveAttribute('disabled');
    expect(getByRole('button')).toHaveClass('disabled');
  });

  it('should call onClick event', (): void => {
    const onClick = jest.fn();
    const { getByRole } = render(<Button onClick={onClick}>{children}</Button>);

    fireEvent.click(getByRole('button'));

    expect(onClick).toHaveBeenCalled();
  });

  it('should not call events if disabled prop is true', (): void => {
    const onClick = jest.fn();
    const { getByRole } = render(<Button onClick={onClick} disabled>{children}</Button>);

    fireEvent.click(getByRole('button'));

    expect(onClick).toHaveBeenCalledTimes(0);
  });

  it('should rerender correctly', (): void => {
    const loading = 'Loading...';
    const { rerender, getByRole } = render(<Button>{children}</Button>);

    expect(getByRole('button')).toHaveTextContent(children);

    rerender(<Button disabled>{loading}</Button>);

    expect(getByRole('button')).toHaveTextContent(loading);
    expect(getByRole('button')).toHaveAttribute('disabled');
  });

  it('should render Link if to prop is true', (): void => {
    const { getByRole } = renderWithProviders(<Button to="/">{children}</Button>);

    expect(getByRole('link')).toHaveTextContent(children);
  });

  it('should have correct appearence if appearence prop is undefined', (): void => {
    const { getByRole } = render(<Button>{children}</Button>);

    expect(getByRole('button')).toHaveClass('primary');
  });

  it('should have correct appearence if appearence prop is "danger"', (): void => {
    const { getByRole } = render(<Button appearence="danger">{children}</Button>);

    expect(getByRole('button')).toHaveClass('danger');
  });

  it('should have correct size if size prop is undefined', (): void => {
    const { getByRole } = render(<Button>{children}</Button>);

    expect(getByRole('button')).toHaveClass('medium');
  });

  it('should have correct size if size prop is small', (): void => {
    const { getByRole } = render(<Button size="small">{children}</Button>);

    expect(getByRole('button')).toHaveClass('small');
  });

  it('should render and match the snapshot', (): void => {
    const defaultButton = render(<Button>{children}</Button>);
    const flatButton = render(<Button flat>{children}</Button>);
    const outlinedButton = render(<Button outlined>{children}</Button>);
    const disabledButton = render(<Button disabled>{children}</Button>);
    const LinkButton = renderWithProviders(<Button to="/">{children}</Button>);

    expect(defaultButton.container.firstChild).toMatchSnapshot();
    expect(flatButton.container.firstChild).toMatchSnapshot();
    expect(outlinedButton.container.firstChild).toMatchSnapshot();
    expect(disabledButton.container.firstChild).toMatchSnapshot();
    expect(LinkButton.container.firstChild).toMatchSnapshot();
  });
});
