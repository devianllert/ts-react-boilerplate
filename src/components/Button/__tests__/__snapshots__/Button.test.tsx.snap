// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render and match the snapshot 1`] = `
<button
  class="button primary medium"
  type="button"
>
  Button
</button>
`;

exports[`<Button /> should render and match the snapshot 2`] = `
<button
  class="button primary medium flat"
  type="button"
>
  Button
</button>
`;

exports[`<Button /> should render and match the snapshot 3`] = `
<button
  class="button primary medium outlined"
  type="button"
>
  Button
</button>
`;

exports[`<Button /> should render and match the snapshot 4`] = `
<button
  class="button primary medium disabled"
  disabled=""
  type="button"
>
  Button
</button>
`;

exports[`<Button /> should render and match the snapshot 5`] = `
<a
  class="button primary medium"
  href="/"
>
  Button
</a>
`;
